genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -379,95 +386,167 @@ private ResolvedJob resolve(final JobResolutionContext context) throws GenieJobR
+                            clusters,
+                            jobRequest,
+                            jobId
+                        );
+                    }
+
+                    if (cluster != null) {
+                        break;
+                    }
+                }
+
+                if (cluster != null) {
+                    break;
+                }
             }
 
             if (cluster == null) {
                 this.noClusterSelectedCounter.increment();
-                throw new GenieJobResolutionException("No cluster found matching given criteria");
+                throw new GenieJobResolutionException("No cluster selected given criteria for job " + jobId);
             }
 
-            log.debug("Selected cluster {} for job {}", cluster.getId(), id);
-            MetricsUtils.addSuccessTags(timerTags);
-            return cluster;
+            log.debug("Selected cluster {} for job {}", cluster.getId(), jobId);
+
+            MetricsUtils.addSuccessTags(tags);
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_ID, cluster.getId()));
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_NAME, cluster.getMetadata().getName()));
+            context.setCluster(cluster);
         } catch (final Throwable t) {
-            MetricsUtils.addFailureTagsWithException(timerTags, t);
+            MetricsUtils.addFailureTagsWithException(tags, t);
             if (t instanceof GenieJobResolutionException) {
                 throw (GenieJobResolutionException) t;
             } else {
                 throw new GenieJobResolutionException(t);
             }
         } finally {
             this.registry
-                .timer(SELECT_CLUSTER_TIMER, timerTags)
-                .record(System.nanoTime() - start, TimeUnit.NANOSECONDS);
-        }
-
-    }
-
-    @Deprecated
-    private Command getCommand(
-        final String commandId,
-        final String jobId
-    ) throws GenieJobResolutionException {
-        final long start = System.nanoTime();
-        final Set<Tag> tags = Sets.newHashSet();
-        try {
-            log.info("Selecting command for job {} ", jobId);
-            final Command command = this.persistenceService.getCommand(commandId);
-            log.info("Selected command {} for job {} ", commandId, jobId);
-            MetricsUtils.addSuccessTags(tags);
-            return command;
-        } catch (final Throwable t) {
-            MetricsUtils.addFailureTagsWithException(tags, t);
-            throw new GenieJobResolutionException(t);
-        } finally {
+                .counter(RESOLVE_CLUSTER_CRITERIA_COMBINATION_COUNTER)
+                .increment(criteriaCombinationCount);
             this.registry
-                .timer(SELECT_COMMAND_TIMER, tags)
+                .counter(RESOLVE_CLUSTER_QUERY_COUNTER)
+                .increment(queryCount);
+            tags.add(Tag.of(RESOLVE_CLUSTER_CRITERIA_COMBINATION_TAG, String.valueOf(criteriaCombinationCount)));
```

Comment:
I think the proper way to make this info available (and easy to query) is a distribution/histogram: https://micrometer.io/docs/concepts#_distribution_summaries

Reply:
Cool. I'll take a look at this for future stuff. In this case this logic is deleted later in the PR anyway as we don't have these nested queries.